cdap-watchdog/src/main/java/co/cask/cdap/logging/appender/remote/RemoteLogAppender.java

```diff
@@ -0,0 +1,168 @@
+
+    private RemoteLogPublisher(CConfiguration cConf, int queueSize) {
+      super(queueSize, RetryStrategies.fromConfiguration(cConf, "system.log.process."));
+      this.numPartitions = cConf.getInt(Constants.Logging.NUM_PARTITIONS);
+      this.loggingEventSerializer = new LoggingEventSerializer();
+      this.logPartitionType =
+        LogPartitionType.valueOf(cConf.get(Constants.Logging.LOG_PUBLISH_PARTITION_KEY).toUpperCase());
+    }
+
+    @Override
+    protected Map.Entry<Integer, byte[]> createMessage(LogMessage logMessage) {
+      String partitionKey = logPartitionType.getPartitionKey(logMessage.getLoggingContext());
+      int partition = partition(partitionKey, numPartitions);
+      return new AbstractMap.SimpleEntry<>(partition, loggingEventSerializer.toBytes(logMessage));
+    }
+
+    @Override
+    protected void publish(List<Map.Entry<Integer, byte[]>> logMessages) throws Exception {
+      // Group the log messages by partition and then publish all messages to their respective partitions
+      Map<Integer, List<byte[]>> partitionedMessages = new HashMap<>();
+      for (Map.Entry<Integer, byte[]> logMessage : logMessages) {
+        List<byte[]> messages = partitionedMessages.computeIfAbsent(logMessage.getKey(), k -> new ArrayList<>());
+        messages.add(logMessage.getValue());
+      }
+
+      for (Map.Entry<Integer, List<byte[]>> partition : partitionedMessages.entrySet()) {
+        ByteArrayOutputStream os = new ByteArrayOutputStream();
+        encodeEvents(os, partition.getKey(), partition.getValue());
+        HttpRequest request = remoteClient.requestBuilder(HttpMethod.POST, "process")
+          .addHeader(HttpHeaders.CONTENT_TYPE, "avro/binary")
+          .withBody(ByteBuffer.wrap(os.toByteArray())).build();
+        HttpResponse response = remoteClient.execute(request);
+        // if something went wrong, throw exception to retry
+        if (response.getResponseCode() != HttpURLConnection.HTTP_OK) {
+          throw new IOException(String.format("Could not append logs for partition %s", partition));
+        }
+      }
+    }
+
+    @Override
+    protected void logError(String errorMessage, Exception exception) {
+      // Log using the status manager
+      addError(errorMessage, exception);
+    }
+  }
+
+  private static int partition(Object key, int numPartitions) {
+    return Math.abs(Hashing.md5().hashString(key.toString()).asInt()) % numPartitions;
+  }
+
+  private void encodeEvents(OutputStream os, int partition, List<byte[]> events) throws IOException {
+    GenericRecord record = new GenericData.Record(LogSchema.LogBufferRequest.SCHEMA);
+    record.put("partition", partition);
+    record.put("events", events.stream().map(ByteBuffer::wrap).collect(Collectors.toList()));
+
+    Encoder encoder = EncoderFactory.get().directBinaryEncoder(os, null);
+
+    DatumWriter<GenericRecord> datumWriter = new GenericDatumWriter<>(LogSchema.LogBufferRequest.SCHEMA);
```

Comment:
does this need to be closed? Also, can it be created once and re-used?

Reply:
yea we can reuse it. I dont see a method to close the writer though.